widgets/text/scroll.go

```diff
@@ -0,0 +1,147 @@
+	// (positive) by a page of content. E.g. -1 means up by one page and 2
+	// means down by two pages.
+	scrollPage int
+
+	// first tracks the first line that will be printed.
+	first int
+
+	// state is the state of the scrolling FSM.
+	state rollState
+}
+
+// newScrollTracker returns a new scroll tracker.
+func newScrollTracker(opts *options) *scrollTracker {
+	if opts.rollContent {
+		return &scrollTracker{state: rollToEnd}
+	}
+	return &scrollTracker{state: rollingDisabled}
+}
+
+// upOneLine processes a user request to scroll up by one line.
+func (st *scrollTracker) upOneLine() {
+	st.scroll--
+}
+
+// downOneLine processes a user request to scroll down by one line.
+func (st *scrollTracker) downOneLine() {
+	st.scroll++
+}
+
+// upOnePage processes a user request to scroll up by one page.
+func (st *scrollTracker) upOnePage() {
+	st.scrollPage--
+}
+
+// downOnePage processes a user request to scroll down by one page.
+func (st *scrollTracker) downOnePage() {
+	st.scrollPage++
+}
+
+// doScroll processes any outstanding scroll requests and calculates the
+// resulting first line.
+func (st *scrollTracker) doScroll(lines, height int) int {
+	first := st.first + st.scroll + st.scrollPage*height
+	st.scroll = 0
+	st.scrollPage = 0
+	return normalizeScroll(first, lines, height)
+}
+
+// firstLine returns the number of the first line that should be drawn on a
+// canvas of the specified height if there is the provided number of lines of
+// text.
+func (st *scrollTracker) firstLine(lines, height int) int {
+	// Execute the scrolling FSM.
+	st.state = st.state(st, lines, height)
+	return st.first
+}
+
+// rollState is a state in the scrolling FSM.
+type rollState func(st *scrollTracker, lines, height int) rollState
+
+// rollingDisabled is a state where content rolling was disabled by the
+// configuration of the Text widget.
+func rollingDisabled(st *scrollTracker, lines, height int) rollState {
+	st.first = st.doScroll(lines, height)
+	return rollingDisabled
+}
+
+// rollToEnd is a state in which the last line of the content is always
+// visible. When new content arrives, it is rolled upwards.
+func rollToEnd(st *scrollTracker, lines, height int) rollState {
+	// If the user didn't scroll, just roll the content so that the last line
+	// is visible.
+	if st.scroll == 0 && st.scrollPage == 0 {
+		st.first = normalizeScroll(math.MaxUint32, lines, height)
+		return rollToEnd
+	}
+
+	st.first = st.doScroll(lines, height)
+	if lastLineVisible(st.first, lines, height) {
+		return rollToEnd
+	}
+	return rollingPaused
+}
+
+// rollingPaused is a state in which the user scrolled up and made the last
+// line scroll out of the view, so the content rolling is paused.
+func rollingPaused(st *scrollTracker, lines, height int) rollState {
+	st.first = st.doScroll(lines, height)
+	if lastLineVisible(st.first, lines, height) {
+		return rollToEnd
+	}
+	return rollingPaused
+}
+
+// lastLineVisible returns true if the last line is visible given drawing that
```

Comment:
Could we re-write this comment to clarify what the last line means. Is it the last line in the buffer?

Reply:
Comment rewritten. Please let me know if this doesn't help.